source/agora/consensus/data/UTXO.d

```diff
@@ -141,6 +141,68 @@ public class TestUTXOSet
         }
         return false;
     }
+
+    private UTXO getUTXO (Hash utxo) nothrow @safe
+    {
+        UTXO value;
+        if (!this.findUTXO_(utxo, value))
+            assert(0);
+        return value;
+    }
+
+    public void updateUTXOStorage (const ref Transaction tx, Height height) @safe
+    {
+        import std.algorithm : any;
+
+        // defaults to next block
+        Height unlock_height = Height(height + 1);
+
+        // for payments of frozen transactions, it will melt after 2016 blocks
+        if ((tx.type == TxType.Payment) && tx.inputs.any!(input =>
+                (
+                    (this.getUTXO(input.utxo).type == TxType.Freeze)
+                )
+            )
+        )
+        {
+            unlock_height = Height(height + 2016);
+        }
+
+        foreach (const ref input; tx.inputs)
+        {
+            this.storage.remove(input.utxo);
```

Comment:
I would add an assert here making sure the utxo exists. It will help catch bugs.